homeassistant/components/sensor/folder.py

```diff
@@ -58,21 +66,45 @@ class Folder(Entity):
 
     ICON = 'mdi:folder'
 
-    def __init__(self, folder_path, filter_term):
+    def __init__(self, folder_path, filter_term, hass):
         """Initialize the data object."""
         folder_path = os.path.join(folder_path, '')  # If no trailing / add it
         self._folder_path = folder_path   # Need to check its a valid path
         self._filter_term = filter_term
+        self._hass = hass
         self._number_of_files = None
+        self._files_dict = {}
         self._size = None
         self._name = os.path.split(os.path.split(folder_path)[0])[1]
         self._unit_of_measurement = 'MB'
 
     def update(self):
         """Update the sensor."""
-        files_list = get_files_list(self._folder_path, self._filter_term)
-        self._number_of_files = len(files_list)
-        self._size = get_size(files_list)
+        current_files = get_files_list(self._folder_path, self._filter_term)
+        self._number_of_files = len(current_files)
+        self._size = get_size(current_files)
+
+        previous_files = list(self._files_dict.keys())
+
+        for file_path in current_files:
+            file_name = os.path.split(file_path)[-1]
+            file_mtime = get_timestamp(file_path)
+
+            if file_path not in self._files_dict:
+                self._hass.bus.fire('file_added', {'file': file_name})
+                self._files_dict[file_path] = file_mtime  # Add the entry
+
+            elif (file_path in self._files_dict and  # If exists and modified
+                    self._files_dict[file_path] != file_mtime):
+                        self._hass.bus.fire(
+                            'file_modified', {'file': file_name})
+                        self._files_dict[file_path] = file_mtime  # Reassign
+
+        # Check if any files deleted
+        for file_path in list(set(previous_files) - set(current_files)):
```

Comment:
I don't think you need to wrap this in a list, an iterator should be just fine + save on memory allocation.

Reply:
Also: can't this be done inside the for loop above? Like `if file_path not in previous_files:`

Reply:
yes